### PR TITLE
Update the API & train_step() of RetinaNet to encourage TPU compatible forward passes

### PR DESCRIPTION
**Pull request by LukeWood**
This API does the following:

- disable train time metric evaluation by default
- this is to encourage users to use model.evaluate() or a SideCarEvaluator for metric evaluation, to keep metrics performant
- override compile to support box loss and classification loss, this will give better metric tracking

### CODE REVIEW

File: examples/models/object_detection/retina_net/basic/pascal_voc/train.py [hunk 1]
```diff
@@ -236,15 +233,9 @@ def unpackage_dict(inputs):
...
-loss = keras_cv.losses.ObjectDetectionLoss(
-    classes=20,
-    classification_loss=keras_cv.losses.FocalLoss(from_logits=True, reduction="none"),
-    box_loss=keras_cv.losses.SmoothL1Loss(l1_cutoff=1.0, reduction="none"),
-    reduction="auto",
-)
-
 model.compile(
-    loss=loss,
+    classification_loss=keras_cv.losses.FocalLoss(from_logits=True),
```
Comment (tanzhenyu) [inline examples/models/object_detection/retina_net/basic/pascal_voc/train.py:237]: Nice!!!

File: keras_cv/utils/train.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tensorflow as tf
+
+
+def scale_loss_for_distribution(loss_value):
```
Comment (tanzhenyu) [inline keras_cv/utils/train.py:18]: Use `[tf.nn.scale_regularization_loss](https://www.tensorflow.org/api_docs/python/tf/nn/scale_regularization_loss)`

File: keras_cv/models/object_detection/retina_net/retina_net.py [hunk 3]
```diff
@@ -224,65 +214,173 @@ def call(self, x, training=False):
...
+            )
+
+        if y_pred.shape[-1] != self.classes + 4:
+            raise ValueError(
+                "y_pred should have shape (None, None, classes + 4). "
+                f"Got y_pred.shape={tuple(y_pred.shape)}.  Does your model's `classes` "
+                "parameter match your losses `classes` parameter?"
+            )
+
+        box_labels = y_true[:, :, :4]
```
Comment (tanzhenyu) [inline keras_cv/models/object_detection/retina_net/retina_net.py:300]: I assume we're gonna change this in next PR?

**Review by tanzhenyu — APPROVED** [review]

**Comment by LukeWood** [comment]
Thanks @tanzhenyu , next up BoxMatcher and TargetGather
